ouroboros-network/src/Ouroboros/Network/Protocol/TxSubmission/Server.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE TypeInType          #-}
+module Ouroboros.Network.Protocol.TxSubmission.Server where
+
+import Control.Monad.Catch (MonadThrow (..))
+import Control.Exception (Exception (..))
+import Data.Functor (($>))
+
+import Protocol.Core
+import Ouroboros.Network.Protocol.TxSubmission.Type
+
+data TxSubmissionServer tx err m a = TxSubmissionServer {
```

Comment:
The client has a `Nat` bound but the server does not?

Reply:
Correct; a server can deal with incorrect clients by throwing the `TxSubmissionExidedCapacityError :: TxSubmissionProtocolError`.

Reply:
If we're enforcing a limit on the client in types, why don't we enforce it in types on the server too?

Reply:
This way one gets an explicit protocol violation error, otherwise one will get a parsing error.  This allows to be more suspicious about a node that did this.